pkg/expression/builtin.go

```diff
@@ -1017,3 +1019,46 @@ func (b *baseBuiltinFunc) MemoryUsage() (sum int64) {
 	}
 	return
 }
+
+type builtinFuncCacheItem[T any] struct {
+	ctxID uint64
+	item  T
+}
+
+type builtinFuncCache[T any] struct {
+	sync.Mutex
+	cached atomic.Pointer[builtinFuncCacheItem[T]]
+}
+
+func (c *builtinFuncCache[T]) getCache(ctxID uint64) (v T, ok bool) {
+	if p := c.cached.Load(); p != nil && p.ctxID == ctxID {
+		return p.item, true
+	}
+	return v, false
+}
+
+func (c *builtinFuncCache[T]) getOrInitCache(ctx EvalContext, constructCache func() (T, error)) (T, error) {
+	intest.Assert(constructCache != nil)
+	ctxID := ctx.GetSessionVars().StmtCtx.CtxID()
+	if item, ok := c.getCache(ctxID); ok {
+		return item, nil
+	}
+
+	c.Lock()
```

Comment:
Maybe we can implement it in CAS pattern as it's already an atomic pointer.

```go
	old := c.cached.Load()
	if old.ctxID == ctxID {
		return old.item, nil
	}

	newItem, err := constructCache()
	if err != nil {
		var def T
		return def, err
	}
	newCache := &builtinFuncCacheItem[T]{
		item:  newItem,
		ctxID: ctxID,
	}
	for !c.cached.CompareAndSwap(old, newCache) {
		old := c.cached.Load()
		if old.ctxID == ctxID {
			return old.item, nil
		}
	}

	return newItem, nil
```

But actually, I have no opinion which one is better :thinking:.

Reply:
The `constructCache` seems to have a probability of being called more than once even if ctxID not change (but I think it is also acceptable).